File: ruby/lib/trees.rb
class BinaryTree
  class Node
    attr_accessor :value, :left, :right, :parent

    def initialize(value, left: nil, right: nil, parent: nil)
      @value = value
      @left = left
      @right = right
      @parent = parent
    end

    def depth
      1 + [left&.depth, right&.depth, 0].compact.max
    end

    def to_s
      res = ''

      if left
        res += '(' + left.to_s + ') '
      end

      res += value.to_s

      if right
        res += ' (' + right.to_s + ')'
      end

      res
    end
  end

  attr_accessor :root

  def initialize(root = nil)
    @root = root
  end

  def to_s
    "(#{root.to_s})"
  end

  def depth
    root&.depth
  end
end

File: go/lib/trees.go
package lib

import (
	"bytes"
	"strconv"
)

type BTNode struct {
	Value  int
	Left   *BTNode
	Right  *BTNode
	Parent *BTNode
}

func (n *BTNode) Depth() int {
	subtreeDepth := 0

	if n.Left != nil {
		subtreeDepth = n.Left.Depth()
	}

	if n.Right != nil {
		if n.Right.Depth() > subtreeDepth {
			subtreeDepth = n.Right.Depth()
		}
	}

	return 1 + subtreeDepth
}

func (n *BTNode) ToString() string {
	res := bytes.Buffer{}

	if n.Left != nil {
		res.WriteString("(")
		res.WriteString(n.Left.ToString())
		res.WriteString(") ")
	}

	res.WriteString(strconv.Itoa(n.Value))

	if n.Right != nil {
		res.WriteString(" (")
		res.WriteString(n.Right.ToString())
		res.WriteString(")")
	}

	return res.String()
}

func (n *BTNode) AppendLeft(value int) *BTNode {
	n.Left = &BTNode{Value: value, Parent: n}
	return n.Left
}

func (n *BTNode) AppendRight(value int) *BTNode {
	n.Right = &BTNode{Value: value, Parent: n}
	return n.Right
}

type BinaryTree struct {
	Root *BTNode
}

func (t *BinaryTree) ToString() string {
	res := bytes.Buffer{}

	res.WriteString("(")
	res.WriteString(t.Root.ToString())
	res.WriteString(")")

	return res.String()
}

func (t *BinaryTree) Depth() int {
	if t.Root != nil {
		return t.Root.Depth()
	} else {
		return 0
	}
}
